File: drills/python/solutions/nc44_gas_station__try_each_start.py
def canCompleteCircuit(gas, cost):
    # Drive the whole loop from each start and see whether the tank ever goes
    # negative. O(n^2) -- the definition, and what the single pass replaces.
    diffs = [g - c for g, c in zip(gas, cost)]

    for start in range(len(diffs)):
        tank = 0
        survives = True
        for diff in diffs[start:] + diffs[:start]:
            tank += diff
            if tank < 0:
                survives = False
                break
        if survives:
            return start

    return -1

File: drills/elixir/solutions/nc44_gas_station__try_each_start.exs
defmodule Solution do
  # Drive the whole loop from each start and see whether the tank ever goes
  # negative. O(n^2) -- the definition, and what the single pass replaces.
  def can_complete_circuit(gas, cost) do
    diffs = gas |> Enum.zip(cost) |> Enum.map(fn {g, c} -> g - c end)

    Enum.find(0..(length(diffs) - 1)//1, -1, fn start ->
      survives?(Enum.drop(diffs, start) ++ Enum.take(diffs, start))
    end)
  end

  defp survives?(rotated) do
    Enum.reduce_while(rotated, 0, fn diff, tank ->
      if tank + diff < 0, do: {:halt, :dry}, else: {:cont, tank + diff}
    end) != :dry
  end
end
